---
title: "Codebase"
author: "Al-Sharif Adly, Madaan Uday, Wu Ivy"
date: "2023-12-09"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
knitr::opts_chunk$set(tidy.opts=list(width.cutoff=80), tidy=TRUE)
library(ggplot2)
library(pheatmap)
library(survival)
library("survminer")
library("SummarizedExperiment")
library("TCGAbiolinks")
library("gridExtra")
library("DESeq2")
library("AnnotationDbi")
library("org.Hs.eg.db")
library("pathview")
library("gage")
library("gageData")
```

####Import Data
```{r}
data_clinical_patient = read.csv("data_clinical_patient.txt", sep = '\t', 
                                 header = FALSE)
data_mutations = read.csv("data_mutations.txt", sep = '\t', header = TRUE)
RNAseq = read.csv("RNAseq_KIRC.csv", sep = ',', header = FALSE)
```

####Get Common Patients
```{r}
clinicalPatients = unique(data_clinical_patient[6:nrow(data_clinical_patient),
                                                1])

mutationPatients = data_mutations[1:nrow(data_mutations),17]
mutationPatients = substr(mutationPatients, 1, 12)
mutationPatients = unique(mutationPatients)

RNApatients = RNAseq[1,2:ncol(RNAseq)]
RNApatients = substr(RNApatients, 1, 12)
RNApatients = unique(RNApatients)

commonPatients <- Reduce(intersect, list(clinicalPatients, mutationPatients, 
                                         RNApatients))
```

####Filtering
```{r}
# Filter data mutations to keep the common patient IDs
data_mutations_filtered <- data_mutations[sapply(
  data_mutations$Tumor_Sample_Barcode, function(x) any(sapply(commonPatients, 
                function(y) grepl(y, substr(x, 1, 12)))), USE.NAMES = FALSE), ]

```


```{r}
# Filter RNAseq to keep the common patient IDs and fix column and row names
columns_to_keep <- sapply(RNAseq, function(col) {
  any(sapply(commonPatients, function(match) grepl(match, 
                                                substr(col[1], 1, 12))))})
columns_to_keep[1]=TRUE
RNAseq_filtered <- RNAseq[, columns_to_keep]
rownames(RNAseq_filtered) <- RNAseq_filtered[, 1]
colnames(RNAseq_filtered) <- RNAseq_filtered[1 ,]
RNAseq_filtered <- RNAseq_filtered[,-1]
RNAseq_filtered <- RNAseq_filtered[-1 ,]
RNAseq_filtered <- as.data.frame(sapply(RNAseq_filtered, as.numeric))
rownames(RNAseq_filtered) <- RNAseq$V1[-1]
nontumor_columns <- grep("^.{13}11", colnames(RNAseq_filtered))
RNAseq_filtered <- RNAseq_filtered[, -nontumor_columns]
colnames(RNAseq_filtered) <- substr(colnames(RNAseq_filtered), 1, 12)
unique_patients <- unique(colnames(RNAseq_filtered))
RNAseq_filtered <- RNAseq_filtered[, unique_patients, drop = FALSE]
```
```{r}
#filter clinical data
rows_to_keep = sapply(data_clinical_patient$V1, function(x) 
  any(sapply(commonPatients, function(y) grepl(y, substr(x, 1, 12)))), 
  USE.NAMES = FALSE)
rows_to_keep[1:5]=TRUE
data_clinical_patient_filtered=data_clinical_patient[rows_to_keep,]
```


##Clinical Analysis

####Clinical hypotheses 
```{r}

# insert clinical hypothesis for survival analysis

HypothesisRows = which(data_clinical_patient_filtered$`Genetic Ancestry Label`[6:359] == "AFR" | data_clinical_patient_filtered$`Genetic Ancestry Label`[6:359] == "AFR_ADMIX" )

HypothesisRows = HypothesisRows + 5
clinicalHypothesis = data_clinical_patient_filtered[HypothesisRows,]

mutationRows = which(substr(data_mutations_filtered$Tumor_Sample_Barcode,1,12) %in% clinicalHypothesis$`#Patient Identifier`)
mutationHypothesis = data_mutations_filtered[mutationRows,]

# create a copy of the clinical data using only rows with actual data
clinicalFilteredCrop = data_clinical_patient_filtered[6:359,]
colnames(clinicalFilteredCrop) = data_clinical_patient_filtered[5,]
clinicalFilteredCrop$deceased = clinicalFilteredCrop$OS_STATUS == "1:DECEASED"
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$DAYS_LAST_FOLLOWUP)/30


# identify which patients are in the hypothesis group
clinicalHypothesisIDs = clinicalHypothesis$`#Patient Identifier`
rowsMatchID = which(clinicalFilteredCrop$PATIENT_ID %in% clinicalHypothesisIDs)
clinicalFilteredCrop$HypothesisGroup = numeric(length(clinicalFilteredCrop$PATIENT_ID))
clinicalFilteredCrop$HypothesisGroup[rowsMatchID] = 1

#remove NAs
rows = which(is.na(clinicalFilteredCrop$Months_to_last_follow_up) == FALSE)
clinicalFilteredCrop = clinicalFilteredCrop[rows,]

# make time numeric
clinicalFilteredCrop$OS_MONTHS = as.numeric(clinicalFilteredCrop$OS_MONTHS)
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$Months_to_last_follow_up)

# do the survival analysis plot
clinicalFilteredCrop$overall_survival = ifelse(clinicalFilteredCrop$deceased == TRUE,
                                   clinicalFilteredCrop$OS_MONTHS, clinicalFilteredCrop$Months_to_last_follow_up)
fit = survfit(Surv(overall_survival, deceased) ~ HypothesisGroup , data=clinicalFilteredCrop)
ggsurvplot(fit, data=clinicalFilteredCrop, pval = T, xlab = "Time (Months)", title = "Survival Analysis", legend.title = "Groups:", legend.labs = c("Other", "Hypothesis"))

```

## Mutation Analysis

####Most Mutated Genes
```{r}
hugo <- as.data.frame(table(data_mutations_filtered$Hugo_Symbol))
hugo.ordered <- hugo[order(-hugo$Freq),]
ggplot(data=hugo.ordered[1:15,], aes(x=Var1, y=Freq))+
  geom_col()+
  theme(axis.text.x = element_text(angle = 45,hjust=1))+ 
  scale_x_discrete(limits = hugo.ordered[1:15,]$Var1)+
  labs(x="Gene", y="Frequency", title="Top 15 Mutated Genes")
```

####Oncomat Matrix
```{r}
cnv_events = unique(data_mutations_filtered$Variant_Classification)
cnv_events
oncomat = reshape2::dcast(
  data = data_mutations_filtered,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]
oncomat.ordered <- oncomat[order(-hugo$Freq),]
mat <- oncomat.ordered
mat[mat != "Silent" & mat != ""]=1
mat[mat == "" | mat == "Silent"]=0
mat <- apply(mat, 2 ,as.numeric)
mat <- as.matrix(mat)
rownames(mat)  <-  row.names(oncomat.ordered)
```

####Gene Mutation Survival Analysis
```{r}
# Test gene survival analysis
#pick gene
mutationCluster = mat[which(rownames(mat) == "LRP2"),]

#turn to df
mutationCluster = data.frame(mutationCluster)

# match clustering from mutation to clinical data
mutationCluster$RowNames <- substr(row.names(mutationCluster),1,12)
matched_rows <- match(clinicalFilteredCrop$PATIENT_ID, mutationCluster$RowNames)
clinicalFilteredCrop$HypothesisGroup <- mutationCluster$mutationCluster[matched_rows]


#remove NAs
rows = which(is.na(clinicalFilteredCrop$Months_to_last_follow_up) == FALSE)
clinicalFilteredCrop = clinicalFilteredCrop[rows,]

# make time numeric
clinicalFilteredCrop$OS_MONTHS = as.numeric(clinicalFilteredCrop$OS_MONTHS)
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$Months_to_last_follow_up)

# do the survival analysis plot
clinicalFilteredCrop$overall_survival = ifelse(clinicalFilteredCrop$deceased == TRUE,
                                   clinicalFilteredCrop$OS_MONTHS, clinicalFilteredCrop$Months_to_last_follow_up)


fit = survfit(Surv(overall_survival, deceased) ~ HypothesisGroup , data=clinicalFilteredCrop)
print(fit)
ggsurvplot(fit, data=clinicalFilteredCrop, pval = T, xlab = "Time (Months)", title = "Survival Analysis", legend.title = "Groups:", legend.labs = c("Other", "Hypothesis"))
```


####Heat Map for top 5 Mutated Genes
```{r}
reduce.mat <- mat[1:5,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE)
```

####Survival Analysis
```{r}
cluster <-  as.data.frame(cutree(res$tree_col, k = 2))
colnames(data_clinical_patient_filtered) <- data_clinical_patient_filtered[1, ]
mutation_patients <- as.character(rownames(cluster))
mutation_patients <- substr(mutation_patients, 1, nchar(mutation_patients) - 3)
survival_df <- data_clinical_patient_filtered[data_clinical_patient_filtered$`#Patient Identifier` %in% mutation_patients,
                                              c("#Patient Identifier",
                                                "Overall Survival Status",
                                                "Overall Survival (Months)")]
survival_df$deceased <- survival_df$`Overall Survival Status`== "1:DECEASED"
survival_df$cluster <- cluster[,1]
survival_df$`Overall Survival (Months)` <- as.numeric(survival_df$`Overall Survival (Months)`)
#Searching for 
fit <- survfit(Surv(`Overall Survival (Months)`, deceased) ~ cluster, data = survival_df)
ggsurvplot(fit, data=survival_df, pval = T)
```

##Expression Analysis

####Exploratory Analysis and Visualization
```{r}
#Prepare RNA data for visualization and DE
RNAseq_matrix = as.matrix(RNAseq_filtered)
RNAseq_matrix <- RNAseq_matrix[rowSums(RNAseq_matrix)>1,]
```

```{r}
#Heat map of euclidean distances between Patients
sampleDists = dist(t(RNAseq_matrix),upper = TRUE)
annot_col = data.frame(cluster[,1])
row.names(annot_col) <- rownames(cluster)

sampleDistMatrix = as.matrix( sampleDists )
rownames(sampleDistMatrix) = substr(colnames(RNAseq_filtered), 1, 12)
colnames(sampleDistMatrix) = substr(colnames(RNAseq_filtered), 1, 12)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         cluster_rows=FALSE, show_rownames=FALSE, show_colnames = FALSE, 
         cluster_cols=TRUE,
         annotation_col=annot_col)
```

```{r}
#PCA Plot
pca_res <- prcomp(t(RNAseq_matrix), scale. = TRUE)
score <- pca_res$x

score = as.data.frame(score)
head(score)
score$color <- as.factor(cluster$`cutree(res$tree_col, k = 2)`)


ggplot(score, aes(x=PC1, y=PC2,  color=color)) + 
  geom_point(size = 4)
```

####Differential Expression Analysis
```{r}
colnames(cluster)[colnames(cluster) == "cutree(res$tree_col, k = 2)"] <- "cluster_no."
rownames(cluster) <- substr(rownames(cluster), 1, 12)
order_in_RNAseq <- match(rownames(cluster), colnames(RNAseq_matrix))
RNAseq_matrix_reordered <- RNAseq_matrix[, order_in_RNAseq]
dds = DESeqDataSetFromMatrix(countData=RNAseq_matrix_reordered,
                              colData=cluster,
                              design=~cluster_no.)
dds = DESeq(dds)
```

```{r}
#Results normalization and gene annotation
res <- results(dds)
res.05 <- results(dds, alpha = 0.05)
resSig <- subset(res.05, padj < 0.05)
ntd <- normTransform(dds)
vsd <- vst(dds)
rownames(res) <- substr(rownames(res), 1, 15)
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    column="SYMBOL",
                    keytype="ENSEMBL",
                    multiVals="first")
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    column="ENTREZID",
                    keytype="ENSEMBL",
                    multiVals="first")
res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    column="GENENAME",
                    keytype="ENSEMBL",
                    multiVals="first")
```

```{r}
#Heat map of top 10 up regulated and down regulated genes against clusters
upGenes <- order(resSig$log2FoldChange, decreasing = TRUE)[1:10]
downGenes <- order(resSig$log2FoldChange, decreasing = FALSE)[1:10]
all.genes <- c(upGenes, downGenes)

annot_col = data.frame(cluster$cluster_no.)
row.names(annot_col) <- rownames(cluster)

sampleMatrix <- assay(vsd)[all.genes,]

rownames(sampleMatrix) = substr(rownames(RNAseq_matrix_reordered[all.genes,]), 1, 15)
colnames(sampleMatrix) = colnames(RNAseq_matrix_reordered)
hugo.genes <- as.data.frame(res$symbol)
matching_genes <- match(rownames(sampleMatrix), rownames(hugo.genes))


for(i in 1:nrow(sampleMatrix)){
  rownames(sampleMatrix)[i] <- hugo.genes$`res$symbol`[matching_genes[i]]
}


pheatmap(sampleMatrix , cluster_rows=FALSE, show_rownames=TRUE, show_colnames = FALSE,
         cluster_cols=TRUE, annotation_col=annot_col)
```

####Pathway Analysis
```{r}
#Pathway Analysis Setup
data(kegg.sets.hs)
data(sigmet.idx.hs)
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
#Pathway Analysis
head(keggres$greater)
head(keggres$less)
```
```{r}
#Getting Pathviews: Put the id of the pathway wanted
pathview(gene.data=foldchanges, pathway.id="hsa04972", species="hsa")
```
