---
title: "Untitled"
author: "Uday Madaan"
date: "2023-11-23"
output: html_document
---


```{r}
#import data
data_clinical_patient = read.csv("data_clinical_patient.txt", sep = '\t', 
                                 header = FALSE)
data_mutations = read.csv("data_mutations.txt", sep = '\t', header = TRUE)
RNAseq = read.csv("RNAseq_KIRC.csv", sep = ',', header = FALSE)
```



```{r}
clinicalPatients = unique(data_clinical_patient[6:nrow(data_clinical_patient),
                                                1])

mutationPatients = data_mutations[1:nrow(data_mutations),17]
mutationPatients = substr(mutationPatients, 1, 12)
mutationPatients = unique(mutationPatients)

RNApatients = RNAseq[1,2:ncol(RNAseq)]
RNApatients = substr(RNApatients, 1, 12)
RNApatients = unique(RNApatients)

commonPatients <- Reduce(intersect, list(clinicalPatients, mutationPatients, 
                                         RNApatients))
```


```{r}
# Filter data mutations to keep the common patient IDs
data_mutations_filtered <- data_mutations[sapply(
  data_mutations$Tumor_Sample_Barcode, function(x) any(sapply(commonPatients, 
                function(y) grepl(y, substr(x, 1, 12)))), USE.NAMES = FALSE), ]

```



```{r}
# Filter RNAseq to keep the common patient IDs
columns_to_keep <- sapply(RNAseq, function(col) {
  any(sapply(commonPatients, function(match) grepl(match, 
                                                substr(col[1], 1, 12))))})
columns_to_keep[1]=TRUE
RNAseq_filtered <- RNAseq[, columns_to_keep]
colnames(RNAseq_filtered) <- unlist(RNAseq_filtered[1, ])  # Setting the column names using the first row
RNAseq_filtered <- RNAseq_filtered[-1, ]
rownames(RNAseq_filtered) <- RNAseq_filtered[,1]  # Assigning the values of the first column as row names
RNAseq_filtered <- RNAseq_filtered[, -1]  
```



```{r}
#Filter clinical data
rows_to_keep = sapply(data_clinical_patient$V1, function(x) 
  any(sapply(commonPatients, function(y) grepl(y, substr(x, 1, 12)))), 
  USE.NAMES = FALSE)
rows_to_keep[1:5]=TRUE
data_clinical_patient_filtered=data_clinical_patient[rows_to_keep,]
```


```{r}
# EXPRESSION ANALYSIS
#RNAseq_filtered
```


```{r}
library(ggplot2)
hugo <- as.data.frame(table(data_mutations$Hugo_Symbol))
hugo.ordered <- hugo[order(-hugo$Freq),]
ggplot(data=hugo.ordered[1:15,], aes(x=Var1, y=Freq))+
  geom_col()+
  theme(axis.text.x = element_text(angle = 45,hjust=1))+
  scale_x_discrete(limits = hugo.ordered[1:15,]$Var1)
```


Oncomat Matrix
```{r}
cnv_events = unique(data_mutations$Variant_Classification)
oncomat = reshape2::dcast(
  data = data_mutations,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]
oncomat.ordered <- oncomat[order(-hugo$Freq),]
mat <- oncomat.ordered
mat[mat!=""]=1
mat[mat==""| mat=="Silent"]=0
mat <- apply(mat, 2 ,as.numeric)
mat <- as.matrix(mat)
rownames(mat)  <-  row.names(oncomat.ordered)
```

```{r}
library(pheatmap)
reduce.mat <- mat[1:5,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE)
cluster <-  as.data.frame(cutree(res$tree_col, k = 3))
```


```{r}

# create a copy of the clinical data using only rows with actual data
clinicalFilteredCrop = data_clinical_patient_filtered[6:359,]
colnames(clinicalFilteredCrop) = data_clinical_patient_filtered[5,]
clinicalFilteredCrop$deceased = clinicalFilteredCrop$OS_STATUS == "1:DECEASED"
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$DAYS_LAST_FOLLOWUP)/30

# do the survival analysis plot
clinicalFilteredCrop$overall_survival = ifelse(clinicalFilteredCrop$deceased == "Dead",
                                   clinicalFilteredCrop$OS_MONTHS, clinicalFilteredCrop$Months_to_last_follow_up)

fit = survfit(Surv(overall_survival, deceased) ~ HypothesisGroup , data=clinicalFilteredCrop[indices_to_keep,])
print(fit)
ggsurvplot(fit, data=clinicalFilteredCrop, pval = T, xlab = "Time (Months)", title = "Survival Analysis", legend.title = "Groups:", legend.labs = c("Other", "Hypothesis"))
```

```{r}
# CLINICAL ANALYSIS

# hypothesis males over 60 are more likely to have renal cell cancer
#HypothesisRows = which(data_clinical_patient_filtered$V6[6:359] == "Male" & data_clinical_patient_filtered$V5[6:359] > 60 )
HypothesisRows = which(data_clinical_patient_filtered$V38[6:359] == "AFR" | data_clinical_patient_filtered$V38[6:359] == "AFR_ADMIX" )
HypothesisRows = HypothesisRows + 5
clinicalHypothesis = data_clinical_patient_filtered[HypothesisRows,]

rows = which(substr(mutation_data$Tumor_Sample_Barcode, 1, 12) %in% unique(clinicalHypothesis$V1))

mutationHypothesis = mutation_data[rows,]


# create the mutation matrix and cluster
cnv_events = unique(mutationHypothesis$Variant_Classification)
oncomat = reshape2::dcast(
  data = mutationHypothesis,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
rnames =  oncomat$Hugo_Symbol
rownames(oncomat) = rnames
oncomat <- oncomat[,-1]
mat = oncomat


mat[mat!=""]=1
mat[mat=="" | mat=="Frame_Shift_Del" | mat=="5'UTR"]=0


mat <- apply(mat, 2 ,as.numeric)
row_sums = rowSums(mat)
sorted_indices = order(row_sums, decreasing = TRUE)
mat = mat[sorted_indices, ]
mat <- as.matrix(mat)
rownames(mat)  <- rnames[sorted_indices]

reduce.mat <- mat[1:5,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE, show_rownames = TRUE)
cluster <-  as.data.frame(cutree(res$tree_col, k = 3))




library("survival")
library("survminer")


# create a copy of the clinical data using only rows with actual data
clinicalFilteredCrop = data_clinical_patient_filtered[6:359,]
colnames(clinicalFilteredCrop) = data_clinical_patient_filtered[5,]
clinicalFilteredCrop$deceased = clinicalFilteredCrop$OS_STATUS == "1:DECEASED"
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$DAYS_LAST_FOLLOWUP)/30




# identify which patients are in the hypothesis group
clinicalHypothesisIDs = clinicalHypothesis$V1
rowsMatchID = which(clinicalFilteredCrop$PATIENT_ID %in% clinicalHypothesisIDs)
clinicalFilteredCrop$HypothesisGroup = numeric(length(clinicalFilteredCrop$PATIENT_ID))
clinicalFilteredCrop$HypothesisGroup[rowsMatchID] = 1



# ensure the null sample size is equal to alternative
set.seed(20) 
# Find the indices of 0s and 1s
zero_indices <- which(clinicalFilteredCrop$HypothesisGroup == 0 &  is.na(clinicalFilteredCrop$Months_to_last_follow_up) == FALSE)
one_indices <- which(clinicalFilteredCrop$HypothesisGroup == 1 & is.na(clinicalFilteredCrop$Months_to_last_follow_up) == FALSE)
print(length(zero_indices))
print(length(one_indices))
#keep a certain number of non 
if (length(zero_indices) > length(one_indices)){
random_zero_indices <- sample(zero_indices, length(one_indices), replace = FALSE)
random_one_indices = one_indices
}else if(length(zero_indices) < length(one_indices)){
  random_one_indices <- sample(one_indices, length(zero_indices), replace = FALSE)
  random_zero_indices = zero_indices
}else{
  random_one_indices = one_indices
  random_zero_indices = zero_indices
  
}
# Combine the indices of 0s and 1s to keep
indices_to_keep <- c(random_one_indices, random_zero_indices)


# do the survival analysis plot
clinicalFilteredCrop$overall_survival = ifelse(clinicalFilteredCrop$deceased == "Dead",
                                   clinicalFilteredCrop$OS_MONTHS, clinicalFilteredCrop$Months_to_last_follow_up)

fit = survfit(Surv(overall_survival, deceased) ~ HypothesisGroup , data=clinicalFilteredCrop[indices_to_keep,])
print(fit)
ggsurvplot(fit, data=clinicalFilteredCrop, pval = T, xlab = "Time (Months)", title = "Survival Analysis", legend.title = "Groups:", legend.labs = c("Other", "Hypothesis"))

```




```{r}
# get RNA table based on the hypothesis
RNAseq_filtered_Hypo = RNAseq_filtered[ , which(substr(colnames(RNAseq_filtered), 1, 12) %in% unique(clinicalHypothesisIDs)) ]
metaData = numeric(length(RNAseq_filtered[,1]))
metaData = ifelse(RNAseq_filtered$Gene == unique(clinicalHypothesisIDs), 1, 0)
RNAseq_filteredToEdit = RNAseq_filtered
```


```{r}

RNAseq_filteredToEdit <- RNAseq_filteredToEdit[rowSums(as.numeric(unlist(RNAseq_filteredToEdit)))>1,]
head(head(countData))
```










