---
title: "Untitled"
author: "Uday Madaan"
date: "2023-11-23"
output: html_document
---


```{r}
#import data
data_clinical_patient = read.csv("data_clinical_patient.txt", sep = '\t', 
                                 header = FALSE)
data_mutations = read.csv("data_mutations.txt", sep = '\t', header = TRUE)
RNAseq = read.csv("RNAseq_KIRC.csv", sep = ',', header = FALSE)
```



```{r}
clinicalPatients = unique(data_clinical_patient[6:nrow(data_clinical_patient),
                                                1])

mutationPatients = data_mutations[1:nrow(data_mutations),17]
mutationPatients = substr(mutationPatients, 1, 12)
mutationPatients = unique(mutationPatients)

RNApatients = RNAseq[1,2:ncol(RNAseq)]
RNApatients = substr(RNApatients, 1, 12)
RNApatients = unique(RNApatients)

commonPatients <- Reduce(intersect, list(clinicalPatients, mutationPatients, 
                                         RNApatients))
```


```{r}
# Filter data mutations to keep the common patient IDs
data_mutations_filtered <- data_mutations[sapply(
  data_mutations$Tumor_Sample_Barcode, function(x) any(sapply(commonPatients, 
                function(y) grepl(y, substr(x, 1, 12)))), USE.NAMES = FALSE), ]

```



```{r}
# Filter RNAseq to keep the common patient IDs
columns_to_keep <- sapply(RNAseq, function(col) {
  any(sapply(commonPatients, function(match) grepl(match, 
                                                substr(col[1], 1, 12))))})
columns_to_keep[1]=TRUE
RNAseq_filtered <- RNAseq[, columns_to_keep]
colnames(RNAseq_filtered) <- unlist(RNAseq_filtered[1, ])  # Setting the column names using the first row
RNAseq_filtered <- RNAseq_filtered[-1, ]
rownames(RNAseq_filtered) <- RNAseq_filtered[,1]  # Assigning the values of the first column as row names
RNAseq_filtered <- RNAseq_filtered[, -1]  
```



```{r}
#Filter clinical data
rows_to_keep = sapply(data_clinical_patient$V1, function(x) 
  any(sapply(commonPatients, function(y) grepl(y, substr(x, 1, 12)))), 
  USE.NAMES = FALSE)
rows_to_keep[1:5]=TRUE
data_clinical_patient_filtered=data_clinical_patient[rows_to_keep,]
```


```{r}
# EXPRESSION ANALYSIS
#RNAseq_filtered
```


```{r}
library(ggplot2)
hugo <- as.data.frame(table(data_mutations_filtered$Hugo_Symbol))
hugo.ordered <- hugo[order(-hugo$Freq),]
ggplot(data=hugo.ordered[1:15,], aes(x=Var1, y=Freq))+
  geom_col()+
  theme(axis.text.x = element_text(angle = 45,hjust=1))+
  scale_x_discrete(limits = hugo.ordered[1:15,]$Var1)
```

```{r}
mat20 = matrix(0, nrow = length(data_mutations_filtered$Hugo_Symbol), ncol = 2)
mat20[,1] = data_mutations_filtered$Hugo_Symbol

# get the number of people with this gene mutated that are dead
for (i in 1:length(data_mutations_filtered$Hugo_Symbol)){
 rows =  which(data_mutations_filtered$Hugo_Symbol == data_mutations_filtered$Hugo_Symbol[i])
 
 names = substr(data_mutations_filtered$Tumor_Sample_Barcode[rows],1,12)
 
 rows = which(data_clinical_patient_filtered[6:359,1] %in% names & data_clinical_patient_filtered$`Overall Survival Status` == "1:DECEASED") + 5
 mat20[i,2] = length(rows)
}


```

```{r}
mat20 <- mat20[!duplicated(mat20[, 1]), ]
mat20 = data.frame(mat20)
mat20[,2] = as.numeric(mat20[,2])
mat20 <- mat20[order(mat20[, 2]), ]
```




Oncomat Matrix
```{r}
cnv_events = unique(data_mutations_filtered$Variant_Classification)
oncomat = reshape2::dcast(
  data = data_mutations_filtered,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]
oncomat.ordered <- oncomat[order(-hugo$Freq),]
mat <- oncomat.ordered
mat[mat=="Missense_Mutation" ]="1"
mat[mat != "1"] = 0

mat <- as.matrix(mat)
mat <- apply(mat, 2 ,as.numeric)

rownames(mat)  <-  row.names(oncomat.ordered)
```


```{r}

row_sums <- rowSums(mat)
sorted_indices <- order(row_sums, decreasing = TRUE)

# Reorder the matrix rows based on the sorted indices
matSort <- mat[sorted_indices, ]
matSort2 =  rowSums(matSort)

num = 10
barplot(t(matSort2[1:num]), beside = TRUE, legend.text = rownames(matSort2[1:num]),
        main = "Number of Patients with a Mutation in Each Gene", xlab = "Genes", ylab = "Number of Patients")
```

```{r}

reduce.mat <- matSort[1:5,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE)

cluster <-  as.data.frame(cutree(res$tree_col, k = 18))




colnames(data_clinical_patient_filtered) <- data_clinical_patient_filtered[1, ]
mutation_patients <- as.character(rownames(cluster))
mutation_patients <- substr(mutation_patients, 1, nchar(mutation_patients) - 3)
survival_df <- data_clinical_patient_filtered[data_clinical_patient_filtered$`#Patient Identifier` %in% mutation_patients,
                                              c("#Patient Identifier",
                                                "Overall Survival Status",
                                                "Overall Survival (Months)")]
survival_df$deceased <- survival_df$`Overall Survival Status`== "1:DECEASED"
survival_df$cluster <- cluster[,1]
survival_df$`Overall Survival (Months)` <- as.numeric(survival_df$`Overall Survival (Months)`)
fit <- survfit(Surv(`Overall Survival (Months)`, deceased) ~ cluster, data = survival_df)
ggsurvplot(fit, data=survival_df, pval = T)
```




```{r}
# CLINICAL ANALYSIS

# insert clinical hypothesis for survival analysis
HypothesisRows = which(data_clinical_patient_filtered$`Genetic Ancestry Label`[6:359] == "AFR" | data_clinical_patient_filtered$`Genetic Ancestry Label`[6:359] == "AFR_ADMIX" )

HypothesisRows = HypothesisRows + 5
clinicalHypothesis = data_clinical_patient_filtered[HypothesisRows,]

mutationRows = which(substr(data_mutations_filtered$Tumor_Sample_Barcode,1,12) %in% clinicalHypothesis$`#Patient Identifier`)
mutationHypothesis = data_mutations_filtered[mutationRows,]

# create the mutation matrix and cluster
cnv_events = unique(mutationHypothesis$Variant_Classification)
oncomat = reshape2::dcast(
  data = mutationHypothesis,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
rnames =  oncomat$Hugo_Symbol
rownames(oncomat) = rnames
oncomat <- oncomat[,-1]
mat = oncomat


mat[mat!=""]=1
mat[mat=="" | mat=="Silent"]=0


mat <- apply(mat, 2 ,as.numeric)
row_sums = rowSums(mat)
sorted_indices = order(row_sums, decreasing = TRUE)
mat = mat[sorted_indices, ]
mat <- as.matrix(mat)
rownames(mat)  <- rnames[sorted_indices]

reduce.mat <- mat[1:5,]
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE, show_rownames = TRUE)
cluster <-  as.data.frame(cutree(res$tree_col, k = 3))




library("survival")
library("survminer")


# create a copy of the clinical data using only rows with actual data
clinicalFilteredCrop = data_clinical_patient_filtered[6:359,]
colnames(clinicalFilteredCrop) = data_clinical_patient_filtered[5,]
clinicalFilteredCrop$deceased = clinicalFilteredCrop$OS_STATUS == "1:DECEASED"
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$DAYS_LAST_FOLLOWUP)/30


# identify which patients are in the hypothesis group
clinicalHypothesisIDs = clinicalHypothesis$`#Patient Identifier`
rowsMatchID = which(clinicalFilteredCrop$PATIENT_ID %in% clinicalHypothesisIDs)
clinicalFilteredCrop$HypothesisGroup = numeric(length(clinicalFilteredCrop$PATIENT_ID))
clinicalFilteredCrop$HypothesisGroup[rowsMatchID] = 1

#remove NAs
rows = which(is.na(clinicalFilteredCrop$Months_to_last_follow_up) == FALSE)
clinicalFilteredCrop = clinicalFilteredCrop[rows,]

# make time numeric
clinicalFilteredCrop$OS_MONTHS = as.numeric(clinicalFilteredCrop$OS_MONTHS)
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$Months_to_last_follow_up)

# do the survival analysis plot
clinicalFilteredCrop$overall_survival = ifelse(clinicalFilteredCrop$deceased == TRUE,
                                   clinicalFilteredCrop$OS_MONTHS, clinicalFilteredCrop$Months_to_last_follow_up)





fit = survfit(Surv(overall_survival, deceased) ~ HypothesisGroup , data=clinicalFilteredCrop)
print(fit)
ggsurvplot(fit, data=clinicalFilteredCrop, pval = T, xlab = "Time (Months)", title = "Survival Analysis", legend.title = "Groups:", legend.labs = c("Other", "Hypothesis"))

```


```{r}
# make the normal mutation matrix
cnv_events = unique(data_mutations_filtered$Variant_Classification)
oncomat = reshape2::dcast(
  data = data_mutations_filtered,
  formula = Hugo_Symbol ~ Tumor_Sample_Barcode,
  fun.aggregate = function(x, cnv = cnv_events) {
    x = as.character(x) # >= 2 same/distinct variant classification = Multi_Hit
    xad = x[x %in% cnv]
    xvc = x[!x %in% cnv]
    
    if (length(xvc) > 0) {
      xvc = ifelse(test = length(xvc) > 1,
                   yes = 'Multi_Hit',
                   no = xvc)
    }
    
    x = ifelse(
      test = length(xad) > 0,
      yes = paste(xad, xvc, sep = ';'),
      no = xvc
    )
    x = gsub(pattern = ';$',
             replacement = '',
             x = x)
    x = gsub(pattern = '^;',
             replacement = '',
             x = x)
    return(x)
  },
  value.var = 'Variant_Classification',
  fill = '',
  drop = FALSE
)
rownames(oncomat) = oncomat$Hugo_Symbol
oncomat <- oncomat[,-1]
oncomat.ordered <- oncomat[order(-hugo$Freq),]
mat <- oncomat.ordered
mat[mat=="Missense_Mutation" ]="1"
mat[mat != "1"] = 0

mat <- as.matrix(mat)
mat <- apply(mat, 2 ,as.numeric)

rownames(mat)  <-  row.names(oncomat.ordered)


# Test gene survival analysis
#pick gene
mutationCluster = mat[which(rownames(mat) == "LRP2"),]

#turn to df
mutationCluster = data.frame(mutationCluster)

# match clustering from mutation to clinical data
mutationCluster$RowNames <- substr(row.names(mutationCluster),1,12)
matched_rows <- match(clinicalFilteredCrop$PATIENT_ID, mutationCluster$RowNames)
clinicalFilteredCrop$HypothesisGroup <- mutationCluster$mutationCluster[matched_rows]


#remove NAs
rows = which(is.na(clinicalFilteredCrop$Months_to_last_follow_up) == FALSE)
clinicalFilteredCrop = clinicalFilteredCrop[rows,]

# make time numeric
clinicalFilteredCrop$OS_MONTHS = as.numeric(clinicalFilteredCrop$OS_MONTHS)
clinicalFilteredCrop$Months_to_last_follow_up = as.numeric(clinicalFilteredCrop$Months_to_last_follow_up)

# do the survival analysis plot
clinicalFilteredCrop$overall_survival = ifelse(clinicalFilteredCrop$deceased == TRUE,
                                   clinicalFilteredCrop$OS_MONTHS, clinicalFilteredCrop$Months_to_last_follow_up)


fit = survfit(Surv(overall_survival, deceased) ~ HypothesisGroup , data=clinicalFilteredCrop)
print(fit)
ggsurvplot(fit, data=clinicalFilteredCrop, pval = T, xlab = "Time (Months)", title = "Survival Analysis", legend.title = "Groups:", legend.labs = c("Other", "Hypothesis"))

```


```{r}
cluster <-  as.data.frame(cutree(res$tree_col, k = 15))
colnames(data_clinical_patient_filtered) <- data_clinical_patient_filtered[1, ]
mutation_patients <- as.character(rownames(cluster))
mutation_patients <- substr(mutation_patients, 1, nchar(mutation_patients) - 3)
survival_df <- data_clinical_patient_filtered[data_clinical_patient_filtered$`#Patient Identifier` %in% mutation_patients,
                                              c("#Patient Identifier",
                                                "Overall Survival Status",
                                                "Overall Survival (Months)")]
survival_df$deceased <- survival_df$`Overall Survival Status`== "1:DECEASED"
survival_df$cluster <- cluster[,1]
survival_df$`Overall Survival (Months)` <- as.numeric(survival_df$`Overall Survival (Months)`)
fit <- survfit(Surv(`Overall Survival (Months)`, deceased) ~ cluster, data = survival_df)
ggsurvplot(fit, data=survival_df, pval = T)
```


```{r}
# get RNA table based on the hypothesis
RNAseq_filtered_Hypo = RNAseq_filtered[ , which(substr(colnames(RNAseq_filtered), 1, 12) %in% unique(clinicalHypothesisIDs)) ]
metaData = numeric(length(RNAseq_filtered[,1]))
metaData = ifelse(RNAseq_filtered$Gene == unique(clinicalHypothesisIDs), 1, 0)
RNAseq_filteredToEdit = RNAseq_filtered
```


```{r}

RNAseq_filteredToEdit = RNAseq_filtered
head(head(RNAseq_filteredToEdit))


sampleDists = dist(t(RNAseq_filteredToEdit), upper = TRUE)
sampleDists
```


```{r}
sampleDists[is.na(sampleDists)] = 0
sampleDists[is.infinite(sampleDists)] <- 0
annot_col = cluster
row.names(annot_col) = substr(rownames(cluster), 1, 12)

sampleDistMatrix = as.matrix( sampleDists )
rownames(sampleDistMatrix) = substr(colnames(RNAseq_filteredToEdit),1,12)
colnames(sampleDistMatrix) = substr(colnames(RNAseq_filteredToEdit),1,12)

pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, annotation_col = annot_col)


```


```{r}
myMat = as.matrix(RNAseq_filteredToEdit)
myMat = matrix(as.numeric(myMat), nrow = nrow(myMat))
pca_res <- prcomp(t(myMat))

score <- pca_res$x


score = as.data.frame(score)
```


```{r}

clusterVector = cluster$`cutree(res$tree_col, k = 18)`


score$color <- as.factor(filtered_cluster)


ggplot(score, aes(x=PC1, y=PC2,  color=color)) + 
  geom_point(size = 4)
```






